pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go

```diff
@@ -78,186 +81,128 @@ type AtlasDatabaseUserReconciler struct {
 // +kubebuilder:rbac:groups="",namespace=default,resources=events,verbs=create;patch
 
 func (r *AtlasDatabaseUserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := r.Log.With("atlasdatabaseuser", req.NamespacedName)
+	atlasDatabaseUser := &akov2.AtlasDatabaseUser{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, atlasDatabaseUser)
+	objectNotFound := err != nil && apiErrors.IsNotFound(err)
+	failedToRetrieve := err != nil && !objectNotFound
 
-	databaseUser := &akov2.AtlasDatabaseUser{}
-	result := customresource.PrepareResource(ctx, r.Client, req, databaseUser, log)
-	if !result.IsOk() {
-		return result.ReconcileResult(), nil
+	switch {
+	case failedToRetrieve:
+		return r.fail(req, err), nil
+	case objectNotFound:
+		return r.notFound(req), nil
 	}
```

Comment:
Rethink the `workflow.Result` has been a topic the recently poped up in my discussions with @s-urbaniak.
Having a "helper" function returning it is actually not helping us with error handling or solving the shallow errors problem. Here specifically I preferred to have the code local as it is as simple as calling `client.Get()` and maybe encapsulate it as repeating the process for other controllers.
I do agree it's a piece of code that's going to be repeated on all controllers and encapsulate it would make a lot of sense but it's about baby steps. For example, this could be encapsulated in the next controller turning into state machine and replaced it with no or low impact.